gq/gq_test.go

```diff
@@ -33,10 +34,91 @@ func TestProveVerify(t *testing.T) {
 
 	ok := signerVerifier.VerifyJWT(gqToken)
 	if !ok {
-		t.Fatal("couldn't verify signature we just made")
+		t.Fatal("signature verification failed")
 	}
 }
 
+func TestVerifyModifiedIdPayload(t *testing.T) {
+	oidcPrivKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	oidcPubKey := &oidcPrivKey.PublicKey
+
+	idToken, err := createOIDCToken(oidcPrivKey, "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// modify the ID Token payload to detect IdP signature invalidity via GQ verify
+	err = modifyTokenPayload(idToken, "fail")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	signerVerifier := gq.NewSignerVerifier(oidcPubKey, 256)
+	gqToken, err := signerVerifier.SignJWT(idToken)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ok := signerVerifier.VerifyJWT(gqToken)
+	if ok {
+		t.Fatal("signature verification passed for invalid payload")
+	}
+}
+
+func TestVerifyModifiedGqPayload(t *testing.T) {
+	oidcPrivKey, err := rsa.GenerateKey(rand.Reader, 2048)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	oidcPubKey := &oidcPrivKey.PublicKey
+
+	idToken, err := createOIDCToken(oidcPrivKey, "test")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	signerVerifier := gq.NewSignerVerifier(oidcPubKey, 256)
+	gqToken, err := signerVerifier.SignJWT(idToken)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	// modify the ID Token payload to detect GQ signature invalidity
+	err = modifyTokenPayload(gqToken, "fail")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ok := signerVerifier.VerifyJWT(gqToken)
+	if ok {
+		t.Fatal("signature verification passed for invalid payload")
+	}
+}
+
+func modifyTokenPayload(token []byte, audience string) error {
+	_, payload, _, err := jws.SplitCompact(token)
+	if err != nil {
+		return err
+	}
+	newPayload := map[string]any{
+		"sub": "1",
+		"iss": "test",
+		"aud": audience,
+		"iat": time.Now().Unix(),
+	}
+	modifiedPayload, err := json.Marshal(newPayload)
+	if err != nil {
+		return err
+	}
+	copy(payload, util.Base64EncodeForJWT(modifiedPayload))
```

Comment:
This is a bit sketchy I think - it will result in an invalid token if the old payload and new payload are not the exact same length, right? I think we should have `modifyTokenPayload` return a new payload rather than modify the existing one.

Reply:
agreed, refactored to create and return a new token rather than modify existing in [275ec15](https://github.com/openpubkey/openpubkey/pull/46/commits/275ec151a15c2144ac4f9e81c38c2341d2b2f9e8)